article/_posts/2020-09-25-Phasic-Policy-Gradient.md

```diff
@@ -0,0 +1,103 @@
+
+To try and have the best of both worlds, the authors propose the _Phasic Policy Gradient_ algorithm, where feature sharing is kept but the two networks training are decoupled.
+
+# Methods
+
+![](/article/images/phasic-pg/fig1.jpeg)
+
+To decouple training while also enabling feature sharing, a novel actor-critic architecture is proposed where the actor and critic use separate networks, while also adding a critic head to the actor network. Figure 1 highlights the proposed architecture. 
+
+To train the networks, the learning process is split into two phases: The policy phase and the auxiliary phase. During the policy phase, the actor is trained using the standard PPO[^1] clipped lost:
+
+![](/article/images/phasic-pg/eq1.jpeg)
+
+where $$r_t(\theta) = \frac{\pi_\theta(a_t{\mid}s_t)}{\pi_{\theta_{old}}(a_t{\mid}s_t)}$$. During the same phase, the critic is also trained using standard $$L_2$$ loss
+
+![](/article/images/phasic-pg/eq2.jpeg)
+
+where $$\hat{V}^{targ}_t$$ and $$\hat{A}_t$$ are computed with GAE[^2]. The _policy_ phase so far is on par with previous works. The novelty comes from the _auxiliary_ phase, where the policy is again optimized, but this time with the following joint loss:
+
+![](/article/images/phasic-pg/eq3.jpeg)
+
+which includes an auxiliary loss $$L^{aux}$$ component for the auxiliary critic head on the actor network, as well as a behavioral cloning loss component to preserve the original policy. $$\pi_{\theta_{old}}$$ is the policy right before the auxiliary phase begins.
+
+$$L^{aux}$$ is again a $$L_2$$ loss
+
+![](/article/images/phasic-pg/eq4.jpeg)
+
+where $$\hat{V}^{targ}_t$$ are the targets computed during the policy phase and $$V_{\theta_\pi}$$ is the critic head shown in Figure 1.
+
+The whole algorithm is shown below:
+
+![](/article/images/phasic-pg/alg1.jpeg)
+
+## Data
+
+The algorithm was tested on the Procgen Benchmark suite[^3].
+
+# Results
+
+## Comparison to PPO
+
+![](/article/images/phasic-pg/res1.jpeg)
+
+We can see that the algorithm learns faster as well as better than PPO on all environments.
+
+## Sample reuse
+
+![](/article/images/phasic-pg/res2.jpeg)
+
+Reusing the same samples for multiple training epochs for the policy does not seem to improve performance.
+
+![](/article/images/phasic-pg/res3.jpeg)
+
+> We find that training with additional auxiliary epochs is generally beneficial, with performance tapering off around 6 auxiliary epochs.
+
+![](/article/images/phasic-pg/res4.jpeg)
+
+> It is clear that performance suffers when we perform auxiliary phases too frequently. We conjecture that each auxiliary phase interferes with policy optimization and that performing frequent auxiliary phases exacerbates this effect.
+
+## Single network vs. multiple networks
+
+![](/article/images/phasic-pg/res5.jpeg)
```

Comment:
How is the `when necessary` part determined?

Reply:
Because the algorithm is modified to only use one network instead of two, loss functions that were ran on the critic network (for example) that wouldn't influence the actor network now would influence it if the gradients were not detached. I'll clarify.